### PR TITLE
Replace Cython kernels with Numba

### PR DESCRIPTION
**Pull request by JSKenyon**
As the title suggests, all the kernels have been moved from Cython to Numba. I have also taken pains to make them compatible with both Python 2 and 3. A side effect of this change is that there should no longer be a painfully long build step. Nothing has changed in the interface, so all existing parsets should continue to work as expected.

@o-smirnov, @bennahugo and @IanHeywood, as power users I would appreciate it if you could use and attempt to break this branch. 

@bennahugo, I would appreciate it if you could run the tests as I cannot run them locally.

### Known issues
1. Currently, I have not specified function signatures. I suspect this makes lazy compilation slow if the inputs to the first run are large. This means the first call to a Numba function will be slow.
2. Threading is implemented but I was struggling to determine if it works correctly inside multiprocessing as my test cases are not large enough. Power users, please check if this works on relevant problems.
3. I am currently abusing the old threading setup in the background. I did this to avoid stripping out all the old OMP code, as I was worried that I would break something. Once the PR is merged, I intend to make a clean up PR in which I strip out the unnecessary code.
4. The docs are now out of date - I will fix them in a separate PR, as I believe that some fiddling will be necessary.

### CODE REVIEW

File: cubical/kernels/chain.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        aa, ab = all_bls[ibl,0], all_bls[ibl,1]
+        for i in range(n_mod):
+            for t in range(n_tim):
+                bt = broadcast_times[t]
+                for f in range(n_fre):
+                    bf = broadcast_freqs[f]
+                    for d in range(n_dir):
+                        bd = broadcast_dirs[d]
+
+                        jh00 = jh[d,i,t,f,aa,ab,0,0]
```
Comment (bennahugo) [inline cubical/kernels/chain.py]: Isn't this loop the wrong way around from a cache coherency perspective - i.e. d on the far right instead?

Comment (o-smirnov) [reply cubical/kernels/chain.py]: When I benchmarked the original cython kernels, I found some really surprising and non-intuitive results w.r.t. loop ordering. Is this order here inherited from cython?

Comment (JSKenyon) [reply cubical/kernels/chain.py]: Yes - I recreated the ordering you found best with Cython. It seemed best with Numba too. So the underlying memory ordering is consistent with the loop ordering even though the "native" ordering remains [d,m,t,f,a,a,c,c] or a variation thereof.

**Comment by JSKenyon** [comment]
@o-smirnov @bennahugo All MS based tests appear to pass under both Py2 and Py3. I am happy with this at this point. 

@o-smirnov @bennahugo @IanHeywood Please let me know if you manage to use this branch successfully, particularly on large data sets here using multiprocessing exclusively is impossible. I still need some confirmation of the threading behaviour.

**Comment by JSKenyon** [comment]
For anyone giving this a bash and requiring Montblanc functionality, from the numba_kernels branch you can run:

`pip install -e ~/CubiCal/".[lsm-support]"`

**Comment by JSKenyon** [comment]
I believe that this is ready to go in. @o-smirnov would you mind glancing at it?

**Comment by bennahugo** [comment]
ok to test

**Pull request merged**